sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/MessageFlux.java

```diff
@@ -0,0 +1,971 @@
+import java.util.function.BiFunction;
+
+import static com.azure.core.amqp.implementation.ClientConstants.DELIVERY_STATE_KEY;
+import static com.azure.core.amqp.implementation.ClientConstants.ENTITY_PATH_KEY;
+import static com.azure.core.amqp.implementation.ClientConstants.LINK_NAME_KEY;
+import static com.azure.core.amqp.implementation.ClientConstants.DELIVERY_TAG_KEY;
+import static com.azure.core.util.FluxUtil.monoError;
+
+/**
+ * The Flux operator to stream messages reliably from a messaging entity (e.g., Event Hub partition,
+ * Service Bus Queue, Topic) to downstream message subscriber.
+ */
+public final class MessageFlux extends FluxOperator<AmqpReceiveLink, Message> {
+    private static final String MESSAGE_FLUX_KEY = "messageFlux";
+    private final ClientLogger logger;
+    private final int prefetch;
+    private final CreditFlowMode creditFlowMode;
+    private final AmqpRetryPolicy retryPolicy;
+    private volatile BiFunction<String, DeliveryState, Mono<Void>> updateDispositionFunc;
+
+    /**
+     * Create a message-flux to stream messages from a messaging entity to downstream subscriber.
+     *
+     * @param source the upstream source that, upon a request, provide a new receiver connected to the messaging entity.
+     * @param prefetch the number of messages that the operator should prefetch from the messaging entity (for a
+     *                 less chatty network and faster message processing on the client).
+     * @param creditFlowMode the mode indicating how to compute the credit and when to send it to the broker.
+     * @param retryPolicy the retry policy to use to obtain a new receiver upon current receiver termination.
+     */
+    public MessageFlux(Flux<? extends AmqpReceiveLink> source, int prefetch, CreditFlowMode creditFlowMode, AmqpRetryPolicy retryPolicy) {
+        super(source);
+
+        final Map<String, Object> loggingContext = new HashMap<>(1);
+        loggingContext.put(MESSAGE_FLUX_KEY, StringUtil.getRandomString("mf"));
+        this.logger = new ClientLogger(MessageFlux.class, loggingContext);
+
+        if (prefetch < 0) {
+            throw new IllegalArgumentException("prefetch >= 0 required but it was " + prefetch);
+        }
+        this.prefetch = prefetch;
+        this.creditFlowMode = creditFlowMode;
+        this.retryPolicy = Objects.requireNonNull(retryPolicy, "'retryPolicy' cannot be null.");
+        this.updateDispositionFunc = (t, s) -> Mono.error(new IllegalStateException("Cannot disposition as no receive-link is established."));
+    }
+
+    /**
+     * Register the downstream subscriber.
+     *
+     * @param actual the downstream subscriber interested in the published messages and termination.
+     */
+    @Override
+    @SuppressWarnings("unchecked")
+    public void subscribe(CoreSubscriber<? super Message> actual) {
+        source.subscribe(new RecoverableReactorReceiver(this, (CoreSubscriber<Message>) actual, prefetch, creditFlowMode, retryPolicy));
```

Comment:
I was curious why we need the explicit cast to Message rather than keeping `<? super Message>` and keeping that generic downstream

Reply:
You're right. We don't need to cast; I've updated downstream to take the generic one.